client/src/js/components/bhBarcodeScanner.js

```diff
@@ -0,0 +1,69 @@
+angular.module('bhima.components')
+  .component('bhBarcodeScanner', {
+    templateUrl : 'modules/templates/bhBarcodeScanner.html',
+    controller  : bhBarcodeScanner,
+    bindings    : {
+      onScanCallback : '&',
+    },
+  });
+
+bhBarcodeScanner.$inject = [
+  '$timeout', '$window', 'BarcodeService',
+];
+
+function bhBarcodeScanner($timeout, $window, Barcode) {
+  const $ctrl = this;
+
+  // steps in the search space
+  const steps = {
+    AWAIT_READ :  2,
+    AWAIT_HTTP : 4,
+    READ_SUCCESS : 8,
+    READ_ERROR : 16,
+    LOST_FOCUS : 32,
+  };
+
+  $ctrl.$onInit = () => {
+    angular.extend($ctrl, { steps });
+    $ctrl.currentStep = steps.AWAIT_READ;
+    $ctrl.setFocusOnHiddenInput();
+  };
+
+  const setFocusOnHiddenInput = () => {
+    // clear previous values
+    delete $ctrl.barcode;
+
+    // find and focus the input
+    const input = $window.document.getElementById('hidden-barcode-input');
+    input.focus();
+
+    // set view state correctly
+    $ctrl.isResetButtonVisible = false;
+    $ctrl.currentStep = steps.AWAIT_READ;
+  };
+
+  // wrap setFocusOnHiddenInput() in $timeout to trigger $digest
+  $ctrl.setFocusOnHiddenInput = () => {
+    $timeout(setFocusOnHiddenInput);
+  };
+
+  $ctrl.triggerBarcodeRead = () => {
+    $ctrl.currentStep = steps.AWAIT_HTTP;
+
+    Barcode.search($ctrl.barcode)
+      .then(record => {
+        $ctrl.currentStep = steps.READ_SUCCESS;
+        $ctrl.record = record;
+        $ctrl.onScanCallback({ record });
+      })
+      .catch(() => {
+        $ctrl.currentStep = steps.READ_ERROR;
```

Comment:
If the record is not found this code falls in the .catch section, and set the step to an error.

I think if the record is not found the step must be READ_FAILED.

On the server side if we use `db.exec`, on the client we can have : 
```js
if (records.length === 0) {
  $ctrl.currentStep = steps.READ_FAILED;
}
```

and the READ_ERROR must be set only if we have a real error.